### PR TITLE
boards/stm32h7: add support for mcuboot

### PR DESCRIPTION
**Pull request by tito97sp**
Add support for MCUBoot.

## Summary
Two new configurations are added:

- **mcuboot-loader:** mcuboot-loader app used as a bootloader.
mcuboot_loader_main` is used as entry point

- **mcuboot-app:** used as mcuboot agent app.
Needs to be signed manually through `"imgtool sign --pad  --align 4 -v 0 -s auto -H 0x200 --pad-header -S 0xc0000 nuttx.hex nuttx_sign.bin"`

Signed-off-by: Andres Sanchez <tito97_sp@hotmail.com>

### CODE REVIEW

File: boards/arm/stm32h7/nucleo-h743zi/Kconfig [hunk 1]
```diff
@@ -27,4 +27,6 @@ config STM32_ROMFS_IMAGEFILE
 	depends on STM32_ROMFS
 	default "../../../rom.img"
 
+
+
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/Kconfig:31]: revert the change

File: boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig [hunk 2]
```diff
@@ -85,3 +86,4 @@ CONFIG_USART3_SERIAL_CONSOLE=y
 CONFIG_USBHOST=y
 CONFIG_USBHOST_MSC=y
 CONFIG_USBHOST_MSC_NOTIFIER=y
+CONFIG_USER_ENTRYPOINT="nsh_main"
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig:89]: remove, the option change to CONFIG_INIT_ENTRYPOINT at line 30

File: boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig [hunk 3]
```diff
@@ -63,6 +63,7 @@ CONFIG_RR_INTERVAL=200
 CONFIG_SCHED_HPWORK=y
 CONFIG_SCHED_LPWORK=y
 CONFIG_SCHED_WAITPID=y
+CONFIG_SDCLONE_DISABLE=y
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig:66]: why change netnsh/defconfig

File: boards/arm/stm32h7/nucleo-h743zi/scripts/gnu-elf.ld [hunk 4]
```diff
@@ -0,0 +1,126 @@
+/****************************************************************************
+ * boards/arm/stm32h7/nucleo-h743zi/scripts/gnu-elf.ld
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/scripts/gnu-elf.ld:2]: remove, let's use binfmt/libelf/gnu-elf.ld instead

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_bringup.c [hunk 5]
```diff
@@ -297,7 +297,7 @@ int stm32_bringup(void)
   if (ret < 0)
     {
       syslog(LOG_ERR, "ERROR: Failed to initialize wireless driver: %d\n",
-             ret);
+            ret);
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_bringup.c:300]: revert the change

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 6]
```diff
@@ -0,0 +1,279 @@
...
+#if defined(CONFIG_STM32_PROGMEM_OTA_PARTITION)
+  ret = init_ota_partitions();
+  if (ret < 0)
+    {
+      return ret;
+    }
+#else
+  /* Use the FTL layer to wrap the MTD driver as a block driver */
+
+  ret = ftl_initialize(PROGMEM_MTD_MINOR, g_progmem_mtd);
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:247]: call register_mtddriver directly

Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:247]: Hi @xiaoxiang781216 . By calling directly `register_mtddriver()` I think I will have to rebuild all the **stm32_progmem** driver. Is this the best way?

File: boards/arm/stm32h7/nucleo-h743zi/Kconfig (hunk 1 above)
Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/Kconfig]: solved

File: boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig (hunk 2 above)
Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig]: solved

File: boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig (hunk 3 above)
Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig]: solved

File: boards/arm/stm32h7/nucleo-h743zi/scripts/gnu-elf.ld (hunk 4 above)
Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/scripts/gnu-elf.ld]: solved

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_bringup.c (hunk 5 above)
Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_bringup.c]: solved

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c (hunk 6 above)
Comment (tito97sp) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:247]: @xiaoxiang781216 Do you have some reference which I could use as template? (i.e: esp32 drivers or so...)

Comment (xiaoxiang781216) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:247]: > @xiaoxiang781216 Do you have some reference which I could use as template? (i.e: esp32 drivers or so...)

Here is how esp32c3 call `register_mtddriver`:
https://github.com/apache/incubator-nuttx/blob/master/arch/risc-v/src/esp32c3/esp32c3_partition.c#L651

> Hi @xiaoxiang781216 . By calling directly `register_mtddriver()` I think I will have to rebuild all the **stm32_progmem** driver. Is this the best way?

You don't need modify any driver code since register_mtddriver use `mtd_dev_s` too.

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_boot_image.c [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+};
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+static void cleanup_arm_nvic(void);
+#ifdef CONFIG_ARMV7M_SYSTICK
+static void systick_disable(void);
+#endif
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_boot_image.c:62]: Please see [`boards/arm/samv7/common/src/sam_boot_image.c`](https://github.com/apache/incubator-nuttx/pull/5406/files#r798493720)

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 8]
```diff
@@ -0,0 +1,279 @@
...
+static struct mtd_dev_s *progmem_alloc_mtdpart(uint32_t mtd_offset,
+                                                   uint32_t mtd_size);
+static int init_ota_partitions(void);
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static FAR struct mtd_dev_s *g_progmem_mtd;
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:96]:
```suggestion
static struct mtd_dev_s *g_progmem_mtd;
```
Please remove `FAR` for ARM files

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    }
+
+#ifdef CONFIG_BCH
+  char blockdev[18];
+  char chardev[12];
+
+  /* Use the minor number to create device paths */
+
+  snprintf(blockdev, 18, "/dev/mtdblock%d", PROGMEM_MTD_MINOR);
+  snprintf(chardev, 12, "/dev/mtd%d", PROGMEM_MTD_MINOR);
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:262]:
```suggestion
  snprintf(blockdev, sizeof(blockdev), "/dev/mtdblock%d", PROGMEM_MTD_MINOR);
  snprintf(chardev, sizeof(chardev), "/dev/mtd%d", PROGMEM_MTD_MINOR);
```

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+            chardev, ret);
+      return ret;
+    }
+#endif /* CONFIG_BCH */
+#endif
+
+  return ret;
+}
+
+#endif /* HAVE_PROGMEM_CHARDEV */
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:279]:
```suggestion
#endif /* HAVE_PROGMEM_CHARDEV */

```

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+  snprintf(blockdev, 18, "/dev/mtdblock%d", PROGMEM_MTD_MINOR);
+  snprintf(chardev, 12, "/dev/mtd%d", PROGMEM_MTD_MINOR);
+
+  /* Now create a character device on the block device */
+
+  ret = bchdev_register(blockdev, chardev, false);
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: bchdev_register %s failed: %d\n",
+            chardev, ret);
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:270]:
```suggestion
      syslog(LOG_ERR, "ERROR: bchdev_register %s failed: %d\n",
             chardev, ret);
```

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 12]
```diff
@@ -0,0 +1,279 @@
...
+      return ret;
+    }
+#else
+  /* Use the FTL layer to wrap the MTD driver as a block driver */
+
+  ret = ftl_initialize(PROGMEM_MTD_MINOR, g_progmem_mtd);
+  if (ret < 0)
+    {
+      syslog(LOG_ERR, "ERROR: Failed to initialize the FTL layer: %d\n",
+            ret);
```
Comment (pkarashchenko) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:251]:
```suggestion
      syslog(LOG_ERR, "ERROR: Failed to initialize the FTL layer: %d\n",
             ret);
```

File: boards/arm/stm32h7/nucleo-h743zi/Kconfig (hunk 1 above)
Comment (acassis) [inline boards/arm/stm32h7/nucleo-h743zi/Kconfig]: Please remove these extra lines

File: boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig (hunk 3 above)
Comment (acassis) [inline boards/arm/stm32h7/nucleo-h743zi/configs/netnsh/defconfig]: It seems you are using old config, this option was removed:
```
$ git grep SDCLONE_DISABLE
ReleaseNotes:* [#5392](https://github.com/apache/incubator-nuttx/pull/#5392) sched: Remove SDCLONE_DISABLE option and config 
```

**Comment by xiaoxiang781216** [comment]
@tito97sp it's better to squash the last patch to previous patch.

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+  int ret = 0;
+
+  g_progmem_mtd = progmem_initialize();
+  if (g_progmem_mtd == NULL)
+    {
+      ferr("ERROR: Failed to get progmem flash MTD\n");
+      ret = -EIO;
+    }
+
+  ret = MTD_IOCTL(g_progmem_mtd, MTDIOC_GEOMETRY, (unsigned long)&geo);
```
Comment (xiaoxiang781216) [inline boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:246]: remove?

**Review by xiaoxiang781216 — APPROVED** [review]

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_boot_image.c (hunk 7 above)
Comment (pkarashchenko) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_boot_image.c:62]: @xiaoxiang781216 why this was marked as resolved? It was not answered.

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c (hunk 8 above)
Comment (pkarashchenko) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:98]: @xiaoxiang781216 why this was marked as resolved? And we bring FAR back for ARM.

Comment (xiaoxiang781216) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_progmem.c:98]: Fix here: https://github.com/apache/incubator-nuttx/pull/7106

File: boards/arm/stm32h7/nucleo-h743zi/src/stm32_boot_image.c (hunk 7 above)
Comment (xiaoxiang781216) [reply boards/arm/stm32h7/nucleo-h743zi/src/stm32_boot_image.c:62]: Fix here https://github.com/apache/incubator-nuttx/pull/7107
